view.php

```diff
@@ -34,6 +34,7 @@
 $appinstance = new \mod_applaunch\applaunch($cm->instance);
 
 require_login($course, false);
+require_capability('mod/applaunch:view', context_module::instance($cmid));
 
 $PAGE->set_url(new moodle_url('/mod/applaunch/view.php', ['id' => $cmid]));
 
```

Comment:
is there a good reason to buffer the html vs directly echo, and not let it stream if it can below?

Reply:
This function isn't in totara 12

Reply:
I find it better for debugging, and find the pattern of a single exit point a better code style. I hadn't considered that there might be a performance hit.

I'll switch this to multiple echo's if you believe it's better.

Reply:
yeah its a subtle thing but it can make the perceived performance better as the html head is sent to the browser and the browser can start parsing the css and rending the page before the rest comes through